### ISSUE TITLE
Fortigate Route Discovery fails on firmware 7.2.x

### ISSUE THREAD

**Issue by smos** [librenms/librenms#15780]
### The problem

When route discovery runs against a Fortigate unit on firmware 7.2.x (7.0.x is ok) the Route attributes are returned as array items tripping up netmask2cidr()

./discovery.php -h firewall -d -v -m route > route-disco.log

RFC1213 [0;31mError discovering route module for FIREWALL.[0m TypeError: ip2long(): Argument #1 ($ip) must be of type string, array given in /opt/librenms/LibreNMS/Util/IPv4.php:74
Stack trace:
#0 /opt/librenms/LibreNMS/Util/IPv4.php(74): ip2long()
#1 /opt/librenms/includes/discovery/route.inc.php(88): LibreNMS\Util\IPv4::netmask2cidr()
#2 /opt/librenms/includes/discovery/functions.inc.php(164): include('...')
#3 /opt/librenms/discovery.php(108): discover_device()
#4 {main}  
SQL[[1;33minsert into `eventlog` (`reference`, `type`, `datetime`, `severity`, `message`, `username`, `device_id`) values (?, ?, ?, ?, ?, ?, ?) [0;33m[null,"discovery","2024-01-30 08:00:41",5,"Error discovering route module. Check log file for more details.","",133][0m 0.83ms] 
  
ip2long(): Argument #1 ($ip) must be of type string, array given {"exception":"[object] (TypeError(code: 0): ip2long(): Argument #1 ($ip) must be of type string, array given at /opt/librenms/LibreNMS/Util/IPv4.php:74)"} 

Line 88 contains the following code:
$inetCidrRoutePfxLen = IPv4::netmask2cidr($ipRoute['ipRouteMask']); //CONVERT

### Output of ./validate.php

```txt
librenms:/opt/librenms$ ./validate.php 
===========================================
Component | Version
--------- | -------
LibreNMS  | 23.11.0 (2023-11-25T20:17:34+01:00)
DB Schema | 2023_11_04_125846_packages_increase_name_column_length (273)
PHP       | 8.1.22
Python    | 3.10.13
Database  | MariaDB 10.6.12-MariaDB-1:10.6.12+maria~ubu2004
RRDTool   | 1.7.2
SNMP      | 5.9.3
===========================================

[OK]    Installed from the official Docker image; no Composer required
[OK]    Database connection successful
[OK]    Database Schema is current
[OK]    SQL Server meets minimum requirements
[OK]    lower_case_table_names is enabled
[OK]    MySQL engine is optimal
[OK]    Database and column collations are correct
[OK]    Database schema correct
[OK]    MySQl and PHP time match
[OK]    Active pollers found
[OK]    Dispatcher Service is enabled
[OK]    Locks are functional
[OK]    No python wrapper pollers found
[OK]    Redis is unavailable
[WARN]  IPv6 is disabled on your server, you will not be able to add IPv6 devices.
[OK]    rrd_dir is writable
[OK]    rrdtool version ok
[WARN]  Updates are managed through the official Docker image
librenms:/opt/librenms$
```

### What was the last working version of LibreNMS?

n/a

### Anything in the logs that might be useful for us?

```txt
SNMP[[0;36m'/usr/bin/snmpbulkwalk' '-v3' '-l' 'authPriv' '-n' "" '-a' 'SHA' '-A' 'PASS' '-u' 'USER' '-x' 'AES' '-X' 'SECRET' '-OQUsetX' '-m' 'RFC1213-MIB' '-M' '/opt/librenms/mibs:/opt/librenms/mibs/fortinet' 'udp:FIREWALL:161' '.1.3.6.1.2.1.4.21'[0m]
ipRouteDest[0.0.0.0].1 = 0.0.0.0
ipRouteDest[10.0.11.0].1 = 10.0.11.0
ipRouteDest[10.0.11.4].1 = 10.0.11.4
ipRouteDest[10.0.11.10].1 = 10.0.11.10
ipRouteDest[10.0.11.16].1 = 10.0.11.16
ipRouteDest[10.0.11.18].1 = 10.0.11.18
ipRouteDest[10.0.11.20].1 = 10.0.11.20
ipRouteIfIndex[0.0.0.0].1 = 2
ipRouteIfIndex[10.0.11.0].1 = 2
ipRouteIfIndex[10.0.11.4].1 = 2
ipRouteIfIndex[10.0.11.10].1 = 2
ipRouteIfIndex[10.0.11.16].1 = 2
ipRouteIfIndex[10.0.11.18].1 = 2
ipRouteIfIndex[10.0.11.20].1 = 2
ipRouteMetric1[0.0.0.0].1 = 10
ipRouteMetric1[10.0.11.0].1 = 51
ipRouteMetric1[10.0.11.4].1 = 51
ipRouteMetric1[10.0.11.10].1 = 51
ipRouteMetric1[10.0.11.16].1 = 101
ipRouteMetric1[10.0.11.18].1 = 26
ipRouteMetric1[10.0.11.20].1 = 0
ipRouteMetric2[0.0.0.0].1 = 0
ipRouteMetric2[10.0.11.0].1 = 0
ipRouteMetric2[10.0.11.4].1 = 0
ipRouteMetric2[10.0.11.10].1 = 0
ipRouteMetric2[10.0.11.16].1 = 0
ipRouteMetric2[10.0.11.18].1 = 0
ipRouteMetric2[10.0.11.20].1 = 0
ipRouteMetric3[0.0.0.0].1 = 0
ipRouteMetric3[10.0.11.0].1 = 0
ipRouteMetric3[10.0.11.4].1 = 0
ipRouteMetric3[10.0.11.10].1 = 0
ipRouteMetric3[10.0.11.16].1 = 0
ipRouteMetric3[10.0.11.18].1 = 0
ipRouteMetric3[10.0.11.20].1 = 0
ipRouteMetric4[0.0.0.0].1 = 0
ipRouteMetric4[10.0.11.0].1 = 0
ipRouteMetric4[10.0.11.4].1 = 0
ipRouteMetric4[10.0.11.10].1 = 0
ipRouteMetric4[10.0.11.16].1 = 0
ipRouteMetric4[10.0.11.18].1 = 0
ipRouteMetric4[10.0.11.20].1 = 0
ipRouteNextHop[0.0.0.0].1 = 10.0.11.20
ipRouteNextHop[10.0.11.0].1 = 10.0.11.20
ipRouteNextHop[10.0.11.4].1 = 10.0.11.20
ipRouteNextHop[10.0.11.10].1 = 10.0.11.20
ipRouteNextHop[10.0.11.16].1 = 10.0.11.20
ipRouteNextHop[10.0.11.18].1 = 10.0.11.20
ipRouteNextHop[10.0.11.20].1 = 0.0.0.0
ipRouteType[0.0.0.0].1 = 4
ipRouteType[10.0.11.0].1 = 4
ipRouteType[10.0.11.4].1 = 4
ipRouteType[10.0.11.10].1 = 4
ipRouteType[10.0.11.16].1 = 4
ipRouteType[10.0.11.18].1 = 4
ipRouteType[10.0.11.20].1 = 3
ipRouteProto[0.0.0.0].1 = 13
ipRouteProto[10.0.11.0].1 = 13
ipRouteProto[10.0.11.4].1 = 13
ipRouteProto[10.0.11.10].1 = 13
ipRouteProto[10.0.11.16].1 = 13
ipRouteProto[10.0.11.18].1 = 13
ipRouteProto[10.0.11.20].1 = 2
ipRouteAge[0.0.0.0].1 = 0
ipRouteAge[10.0.11.0].1 = 0
ipRouteAge[10.0.11.4].1 = 0
ipRouteAge[10.0.11.10].1 = 0
ipRouteAge[10.0.11.16].1 = 0
ipRouteAge[10.0.11.18].1 = 0
ipRouteAge[10.0.11.20].1 = 0
ipRouteMask[0.0.0.0].1 = 0.0.0.0
ipRouteMask[10.0.11.0].1 = 255.255.255.254
ipRouteMask[10.0.11.4].1 = 255.255.255.254
ipRouteMask[10.0.11.10].1 = 255.255.255.254
ipRouteMask[10.0.11.16].1 = 255.255.255.254
ipRouteMask[10.0.11.18].1 = 255.255.255.254
ipRouteMask[10.0.11.20].1 = 255.255.255.254
ipRouteMetric5[0.0.0.0].1 = 0
ipRouteMetric5[10.0.11.0].1 = 0
ipRouteMetric5[10.0.11.4].1 = 0
ipRouteMetric5[10.0.11.10].1 = 0
ipRouteMetric5[10.0.11.16].1 = 0
ipRouteMetric5[10.0.11.18].1 = 0
ipRouteMetric5[10.0.11.20].1 = 0
ipRouteInfo[0.0.0.0].1 = ccitt.0.0
ipRouteInfo[10.0.11.0].1 = ccitt.0.0
ipRouteInfo[10.0.11.4].1 = ccitt.0.0
ipRouteInfo[10.0.11.10].1 = ccitt.0.0
ipRouteInfo[10.0.11.16].1 = ccitt.0.0
ipRouteInfo[10.0.11.18].1 = ccitt.0.0
ipRouteInfo[10.0.11.20].1 = ccitt.0.0


Routing table:  
array (
  '0.0.0.0' => 
  array (
    'ipRouteDest' => 
    array (
      1 => '0.0.0.0',
    ),
    'ipRouteIfIndex' => 
    array (
      1 => '2',
    ),
    'ipRouteMetric1' => 
    array (
      1 => '10',
    ),
    'ipRouteMetric2' => 
    array (
      1 => '0',
    ),
    'ipRouteMetric3' => 
    array (
      1 => '0',
    ),
    'ipRouteMetric4' => 
    array (
      1 => '0',
    ),
    'ipRouteNextHop' => 
    array (
      1 => '10.0.11.20',
    ),
    'ipRouteType' => 
    array (
      1 => '4',
    ),
    'ipRouteProto' => 
    array (
      1 => '13',
    ),
    'ipRouteAge' => 
    array (
      1 => '0',
    ),
    'ipRouteMask' => 
    array (
      1 => '0.0.0.0',
    ),
    'ipRouteMetric5' => 
    array (
      1 => '0',
    ),
    'ipRouteInfo' => 
    array (
      1 => 'ccitt.0.0',
    ),
  ),
  '10.0.11.0' => 
  array (
    'ipRouteDest' => 
    array (
      1 => '10.0.11.0',
    ),
    'ipRouteIfIndex' => 
    array (
      1 => '2',
    ),
    'ipRouteMetric1' => 
    array (
      1 => '51',
    ),
    'ipRouteMetric2' => 
    array (
      1 => '0',
    ),
    'ipRouteMetric3' => 
    array (
      1 => '0',
    ),
    'ipRouteMetric4' => 
    array (
      1 => '0',
    ),
    'ipRouteNextHop' => 
    array (
      1 => '10.0.11.20',
    ),
    'ipRouteType' => 
    array (
      1 => '4',
    ),
    'ipRouteProto' => 
    array (
      1 => '13',
    ),
    'ipRouteAge' => 
    array (
      1 => '0',
    ),
    'ipRouteMask' => 
    array (
      1 => '255.255.255.254',
    ),
    'ipRouteMetric5' => 
    array (
      1 => '0',
    ),
    'ipRouteInfo' => 
    array (
      1 => 'ccitt.0.0',
    ),
  ),
  '10.0.11.4' => 
  array (
    'ipRouteDest' => 
    array (
      1 => '10.0.11.4',
    ),
    'ipRouteIfIndex' => 
    array (
      1 => '2',
    ),
    'ipRouteMetric1' => 
    array (
      1 => '51',
    ),
    'ipRouteMetric2' => 
    array (
      1 => '0',
    ),
    'ipRouteMetric3' => 
    array (
      1 => '0',
    ),
    'ipRouteMetric4' => 
    array (
      1 => '0',
    ),
    'ipRouteNextHop' => 
    array (
      1 => '10.0.11.20',
    ),
    'ipRouteType' => 
    array (
      1 => '4',
    ),
    'ipRouteProto' => 
    array (
      1 => '13',
    ),
    'ipRouteAge' => 
    array (
      1 => '0',
    ),
    'ipRouteMask' => 
    array (
      1 => '255.255.255.254',
    ),
    'ipRouteMetric5' => 
    array (
      1 => '0',
    ),
    'ipRouteInfo' => 
    array (
      1 => 'ccitt.0.0',
    ),
  ),
```

**Comment by smos** [comment]
There is an open case with TAC and is currently in "Researching"

**Comment by rudybroersma** [comment]
I've also opened a ticket at Fortinet. Ticketnumber #9203141

What FortiNet in all their wisdom has done is add the vDOM ID to certain indexes.
For example, they have added the vDOM is to the BGP4-MIB::bgpPeerTable index. 

I can understand their reasoning, but their implementation is just wrong and breaks MIB compliancy.

For example:
BGP4-MIB::bgpPeerEntry has INDEX { bgpPeerRemoteAddr }
bgpPeerRemoteAddr is:

        bgpPeerRemoteAddr OBJECT-TYPE
            SYNTAX     IpAddress

However.. What FortiNet has done is make the index something like this: 255.255.255.255.1
The .1 suffix is the vDOM ID. This makes the index no longer a valid IpAddress

In my ticket I've asked FortiNet revert this change and implement this in some other way that doesn't break MIB compliancy. I don't have high hopes.

**Comment by rudybroersma** [comment]
ffs. I just got off a 1-hour call with Fortinet regarding other SNMP issues. They have no understanding of SNMP and it's indeed very hard to get the point across.

During the call they also told me that "you should only use Fortinet MIB". I then pointed them to the FortiSwitch datasheet which refers quite the list of MIBs from IETF RFCs. That seemed to have convinced them. But they still don't grasp the basics of SNMP.

Two other tickets of mine regarding SNMP bugs in FortiNet devices have now been escalated to engineering. If that doesn't go anywhere I'll escalate this through our SE. 

Tomorrow morning I have another call scheduled with Fortinet regarding this BGP4-MIB issue.

**Comment by rudybroersma** [comment]
I had a call with Fortinet TAC about this issue to explain to them what's wrong. I just got feedback they created an internal ticket for engineering. Ticket number 1002260.

Still not overly hopeful. The TAC engineer made comments like "we might just choose to ignore the RFC". I cried in despair.

**Comment by rudybroersma** [comment]
"I have confirmation from our developers that the reported issue is matching internal ticket 0883606, that issue will be resolved on 7.2.8 and 7.4.4 with new introduced CLI sintax where you can enable or disable the append-index :

CLI Changes:
config system snmp sysinfo
set append-index {enable | disable }
end 

# FOS now allows customers to enable or disable the INDEX extension, which appends VDOM or interface index in RFC tables.  

7.2.8 GA release is planed to be released next week , please let me know if you want to keep the ticket open till this firmware is available or we can close the ticket ."

**Comment by smos** [comment]
Confirmed fixed in 7.2.8 on 60F.

config system snmp sysinfo
set append-index disable
